src/Lean/Meta/Tactic/LibrarySearch.lean

```diff
@@ -0,0 +1,397 @@
+      let a := a.push (←e.mkSubEntry 0 (name, DeclMod.mp))
+      let a := a.push (←e.mkSubEntry 1 (name, DeclMod.mpr))
+      pure a
+    else
+      pure a
+
+/--
+Candidate finding function that uses strict discrimination tree for resolution.
+-/
+def mkImportFinder : IO CandidateFinder := do
+  let ref ← IO.mkRef none
+  pure fun ty => do
+    let importTree ← (←ref.get).getDM $ do
+      profileitM Exception  "librarySearch launch" (←getOptions) $
+        createImportedEnvironment (←getEnv) (constantsPerTask := constantsPerTask) addImport
+    let (imports, importTree) ← importTree.getMatch ty
+    ref.set importTree
+    pure imports
+
+end IncDiscrTreeFinder
+
+initialize registerTraceClass `Tactic.stdLibrarySearch
+initialize registerTraceClass `Tactic.stdLibrarySearch.lemmas
+
+/-- State for resolving imports -/
+private def LibSearchState := IO.Ref (Option CandidateFinder)
+
+private initialize LibSearchState.default : IO.Ref (Option CandidateFinder) ← do
+  IO.mkRef .none
+
+private instance : Inhabited LibSearchState where
+  default := LibSearchState.default
+
+private initialize ext : EnvExtension LibSearchState ←
+  registerEnvExtension (IO.mkRef .none)
+
+/--
+The preferred candidate finding function.
+-/
+initialize defaultCandidateFinder : IO.Ref CandidateFinder ← unsafe do
+  IO.mkRef (←IncDiscrTreeFinder.mkImportFinder)
+
+/--
+Update the candidate finder used by library search.
+-/
+def setDefaultCandidateFinder (cf : CandidateFinder) : IO Unit :=
+  defaultCandidateFinder.set cf
+
+/--
+Return an action that returns true when  the remaining heartbeats is less
+than the currently remaining heartbeats * leavePercent / 100.
+-/
+def mkHeartbeatCheck (leavePercent : Nat) : MetaM (MetaM Bool) := do
+  let maxHB ← getMaxHeartbeats
+  let hbThreshold := (← getRemainingHeartbeats) * leavePercent / 100
+  -- Return true if we should stop
+  pure $
+    if maxHB = 0 then
+      pure false
+    else do
+      return (← getRemainingHeartbeats) < hbThreshold
+
+private def librarySearchEmoji : Except ε (Option α) → String
+| .error _ => bombEmoji
+| .ok (some _) => crossEmoji
+| .ok none => checkEmoji
+
+/--
+Interleave x y interleaves the elements of x and y until one is empty and then returns
+final elements in other list.
+-/
+def interleaveWith {α β γ} (f : α → γ) (x : Array α) (g : β → γ) (y : Array β) : Array γ :=
+    Id.run do
+  let mut res := Array.mkEmpty (x.size + y.size)
+  let n := min x.size y.size
+  for h : i in [0:n] do
+    have p : i < min x.size y.size := h.2
+    have q : i < x.size := Nat.le_trans p (Nat.min_le_left ..)
+    have r : i < y.size := Nat.le_trans p (Nat.min_le_right ..)
+    res := res.push (f x[i])
+    res := res.push (g y[i])
+  let last :=
+        if x.size > n then
+          (x.extract n x.size).map f
+        else
+          (y.extract n y.size).map g
+  pure $ res ++ last
+
+
+/--
+An exception Id that indicates further speculation on candidate lemmas should stop
```

Comment:
```suggestion
An exception ID that indicates further speculation on candidate lemmas should stop
```
We typically lowercase things in Lean APIS that would otherwise be all-caps, like "JSON" or "HTML", but the English should still use the all-caps